### PR TITLE
Custom fields and add tests convert to auth0-extension-ui and improve unit testing

### PR DESCRIPTION
**Pull request by mostekcm**
This is just for preview

### CODE REVIEW

File: client/components/Users/UserInfo.jsx [hunk 1]
```diff
@@ -35,37 +38,75 @@ export default class UserInfo extends Component {
...
+
+    return value;
   }
 
   render() {
     const { user, error, loading, memberships } = this.props;
-    const currentMemberships = this.getMemberships(memberships);
-    const identity = this.getIdentities(user);
-    const blocked = this.getBlocked(user);
+    const excludeCustomFields = ( this.props.userInfoFields && this.props.userInfoFields.excludeFields ) || [];
```
Comment (sandrinodimattia) [inline client/components/Users/UserInfo.jsx]: This is starting to become a lot of logic to execute for every render. Should be moved to a higher level.

File: server/routes/users.js [hunk 2]
```diff
@@ -25,20 +25,7 @@ export default (storage, scriptManager) => {
...
         user: req.user
       },
-      payload: {
-        email: req.body.email,
-        username: req.body.username,
-        connection: req.body.connection,
-        memberships: req.body.memberships,
-        password: req.body.password
-      },
-      defaultPayload: {
```
Comment (sandrinodimattia) [inline server/routes/users.js]: Why is this removed?

File: server/routes/users.js [hunk 3]
```diff
@@ -85,9 +75,14 @@ export default (storage, scriptManager) => {
   /*
    * Get a single user.
    */
-  api.get('/:id', verifyUserAccess('read:user', scriptManager), (req, res, next) => {
```
Comment (sandrinodimattia) [inline server/routes/users.js]: Why was this changed?

**Comment by Verlic** [comment]
MD5: 93b885adfe0da089cdf634904fd59f71

# NSP - Security Report

_This is an automated security audit of this project. Please do not modify its contents_

Vulnerabilities found: **1**
Affected Modules: **safe-eval**
Summary:

- Sandbox Breakout


## Detailed report
#### Sandbox Breakout

Affected Module: **safe-eval**
Installed version: **0.3.0**
Patched versions: **<0.0.0**
Advisory: https://nodesecurity.io/advisories/337
Path: **auth0-delegated-admin@2.4.4 > safe-eval@0.3.0**

---

File: client/app.jsx [hunk 4]
```diff
@@ -65,7 +66,7 @@ ReactDOM.render(
   document.getElementById('app')
 );
 
-store.dispatch(push('/users'));
+// store.dispatch(push('/users'));
```
Comment (dctoon) [inline client/app.jsx]: Remove?

File: client/components/Users/UsersTable.jsx [hunk 5]
```diff
@@ -9,43 +10,205 @@ import {
...
     users: React.PropTypes.array.isRequired,
-    loading: React.PropTypes.bool.isRequired
+    loading: React.PropTypes.bool.isRequired,
+    userFields: React.PropTypes.array.isRequired,
+    onColumnSort: React.PropTypes.func.isRequired,
+    sortOrder: React.PropTypes.number.isRequired,
+    sortProperty: React.PropTypes.string.isRequired,
+  }
+
+  // shouldComponentUpdate(nextProps) {
```
Comment (dctoon) [inline client/components/Users/UsersTable.jsx]: Remove commented code.

File: server/routes/users.js [hunk 6]
```diff
@@ -8,42 +9,57 @@ import config from '../lib/config';
...
+        request: {
+          user: req.user,
+          originalUser: user
+        },
+        payload: req.body,
+        userFields
+      };
+      return context;
+    })
+    .then(context => scriptManager.execute('create', context));
```
Comment (dctoon) [inline server/routes/users.js]: There is not `create` script anymore, but instead a `write` script should that not be reflected in this code as well? If this is for backwards compatibility, what's the strategy to change this at a later stage?

File: server/routes/users.js (hunk 2 above)
Comment (mostekcm) [reply server/routes/users.js]: It was removed because with custom fields there are more fields that need to be able to be set and it is configured in the settings hook.  We could potentially set this only based on whether or not the userFields array is blank or not.

File: server/routes/users.js (hunk 3 above)
Comment (mostekcm) [reply server/routes/users.js]: It probably should not be... I will look at this.

File: server/routes/users.js (hunk 6 above)
Comment (mostekcm) [reply server/routes/users.js]: There is a way to create an update hook.  Perhaps we need to create an update hook for this one...

Comment (mostekcm) [reply server/routes/users.js]: It could rename the create hook to the write hook.

File: client/components/Users/UserInfo.jsx (hunk 1 above)
Comment (mostekcm) [reply client/components/Users/UserInfo.jsx]: Fixed in next version

File: server/routes/users.js (hunk 3 above)
Comment (mostekcm) [reply server/routes/users.js]: Added back in

File: client/app.jsx (hunk 4 above)
Comment (mostekcm) [reply client/app.jsx]: done

File: client/components/Users/UsersTable.jsx (hunk 5 above)
Comment (mostekcm) [reply client/components/Users/UsersTable.jsx]: done
